agr_literature_service/lit_processing/data_ingest/pubmed_ingest/load_pmc_metadata.py

```diff
@@ -80,11 +82,28 @@ def load_ref_file_metadata_into_db():  # pragma: no cover
         db_session.close()
 
 
-def resolve_displayname_conflict(file_name_with_suffix):
+def resolve_displayname_conflict(ref_file_uniq_filename_set, file_name_with_suffix, reference_id):
+    # this function that appends an increasing number to the base file name until
+    # a unique file name is found:
+    #
+    # We first extract the base name from the input "file_name_with_suffix",
+    # and initialize a counter to 1. We then append the current count value to the
+    # base name to form the new display name.
+    #
+    # If the (reference_id, file_name_with_suffix) tuple is already present in the
+    # "ref_file_uniq_filename_set", we increment the counter, update the display name,
+    # and try again. We repeat this process until a unique file name is found.
 
     file_extension = file_name_with_suffix.split(".")[-1].lower()
-    display_name = file_name_with_suffix.replace("." + file_extension, "") + "_1"
-    return display_name + "." + file_extension
+    base_name = file_name_with_suffix.replace("." + file_extension, "")
```

Comment:
This is the code I added to the function `file_upload_single` in the referencefile_crud:

```
original_name = metadata["display_name"]
        for counter in count(start=1, step=1):
            ref_file_same_name = db.query(ReferencefileModel).filter(
                and_(
                    ReferencefileModel.display_name == metadata["display_name"],
                    ReferencefileModel.file_extension == metadata["file_extension"],
                    ReferencefileModel.reference.has(ReferenceModel.curie == metadata["reference_curie"])
                )
            ).one_or_none()
            if ref_file_same_name is None:
                break
            metadata["display_name"] = f"{original_name}_{counter}"
```

Reply:
Wow. this looks great. Thanks @valearna !! 

In the PMC script, it needs to check the reference_id/display name pair for a long list of files so saving data in memory instead of checking one by one seems more efficient.

Anyway, I will leave the code as is.